# Descripción del proyecto

Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Las hipótesis son suposiciones hechas sobre una población con base en una muestra de datos. Estas suposiciones se prueban utilizando métodos estadísticos para determinar si son verdaderas o falsas. Una prueba de hipótesis consiste en hacer una declaración sobre la población, recopilar datos y usar métodos estadísticos para probar esa declaración.

Siempre que realizamos un análisis, tenemos que formular hipótesis que luego podamos poner a prueba. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

## Descripción del proyecto

En el contexto de este proyecto, probarás una hipótesis relacionada con las preferencias musicales de dos ciudades. Esto se hará mediante el análisis de datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

Esto implicará analizar los datos reales de transmisión de música online para comparar el comportamiento de los usuarios y las usuarias en Springfield y Shelbyville. El proyecto se divide en tres etapas, cada una con objetivos específicos.

En la etapa 1, proporcionarás una descripción general de los datos y anotarás tus observaciones. En la etapa 2, preprocesarás los datos al limpiarlos. Finalmente, en la etapa 3, pondrás a prueba la hipótesis siguiendo los pasos de programación necesarios para probar cada declaración y comentando tus resultados en los bloques apropiados.

Al completar estas etapas, podrás extraer información valiosa de los datos y tomar decisiones basadas en datos.

## Hipótesis

Para este proyecto, recopilamos los requisitos y desarrollamos una hipótesis que debe ser confirmada o negada.

Al probar hipótesis, es importante tener en cuenta que pueden aceptarse por completo, aceptarse parcialmente, rechazarse parcialmente o rechazarse por completo.

Cuando una hipótesis se acepta por completo, significa que los resultados de la prueba respaldan sin ambigüedades la declaración hecha sobre la población.

Si se acepta solo parcialmente, significa que los resultados respaldan la declaración en cierta medida, pero no lo suficiente como para aceptarla por completo.

Por el contrario, si una hipótesis se rechaza por completo, significa que los resultados de la prueba no respaldan la declaración hecha sobre la población.

Por último, una hipótesis también puede ser rechazada parcialmente si los datos indican que es falsa, pero no puedes rechazarla por completo. Cuando se interpretan los resultados de una prueba de hipótesis, es importante tener en cuenta estas diferentes posibilidades.

Aquí está nuestra hipótesis para aceptar o rechazar:

 1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.

## Diccionario de datos

Los datos se almacenan en el archivo **/datasets/music_project_en.csv**



### Descripción de las columnas:

 - __'userID':__ identifica de forma exclusiva a cada usuario o usuaria.
 - __'Track':__ título de la canción.
 - __'artist':__ nombre del artista.
 - __'genre':__ género musical.
 - __'City':__ ciudad del usuario o la usuaria.
 - __'time':__ hora del día en la que se reprodujo la pista (HH:MM:SS).
 - __'Day':__ día de la semana.

## Descripción de los datos

In [1]:
# Importar pandas
import pandas as pd
import warnings

In [2]:
warnings.filterwarnings('ignore')

In [3]:
# Leer el archivo y almacenarlo en df
df = pd.read_csv('datasets/music_project_en.csv')

In [4]:
# Obtener las 10 primeras filas de la tabla df
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

In [5]:
# Obtener la información general sobre nuestros datos
df.dtypes

  userID    object
Track       object
artist      object
genre       object
  City      object
time        object
Day         object
dtype: object

Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. Se debe utilizar snake_case aunque solo es em la columna 'userID pero tambien seria bueno cambiar la columna time para que sea mas especifica ya que puede confundirse con tiempo de repoducion.

In [6]:
print(df.isna().sum())
print('----------------------')
print('Total de valores asuentes: ', df.isna().sum().sum())
print('----------------------')
print('Numero de filas duplicadas:', df.duplicated().sum())

  userID       0
Track       1343
artist      7567
genre       1198
  City         0
time           0
Day            0
dtype: int64
----------------------
Total de valores asuentes:  10108
----------------------
Numero de filas duplicadas: 3826


### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles:

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`
>Los datos son de tipo **object**, los nombres de las columnas son descriptivas por lo que podemos saber que los datos contienen lo suficiente para nuestro analisis

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`
>Tenemos los datos necesarios ya que hay que determinar si la actividad difiere segun el dia de la semana y la ciudad

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`
>Hay 3826 duplicados y 10108 valores ausentes

## Preprocesamiento de datos

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.

### Estilo del encabezado
Muestra los encabezados de la tabla (los nombres de las columnas):

In [7]:
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [8]:
# Bucle en los encabezados poniendo todo en minúsculas
col_lowered = []

for old_names in df.columns:
    name_lowered = old_names.lower()
    col_lowered.append(name_lowered)
    
df.columns = col_lowered
print(df.columns)

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [9]:
# Bucle en los encabezados eliminando los espacios
col_striped = []

for old_names in df.columns:
    name_striped = old_names.strip()
    col_striped.append(name_striped)
    
df.columns = col_striped
print(df.columns)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [10]:
# Cambiar el nombre de la columna "userid"
df.rename(columns={'userid': 'user_id'}, inplace=True)
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Comprueba el resultado. Muestra los encabezados una vez más:

In [11]:
# Comprobar el resultado: la lista de encabezados
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

### Valores ausentes

Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [12]:
# Calcular el número de valores ausentes
print(df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [13]:
replace_nan = ['track', 'artist', 'genre']

# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
for nan in replace_nan:
    df[nan].fillna('unknown', inplace=True)

In [14]:
# Contar valores ausentes
print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


### Duplicados

Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [15]:
# Contar duplicados explícitos
print('Número de duplicados:', df.duplicated().sum())

Número de duplicados: 3826


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [16]:
# Eliminar duplicados explícitos
df = df.drop_duplicates()

Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [17]:
# Comprobar de nuevo si hay duplicados
print('Número de duplicados:', df.duplicated().sum())

Número de duplicados: 0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.

In [18]:
# Inspeccionar los nombres de géneros únicos
print(df['genre'].sort_values().unique())

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [19]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)
    return df

wrong_genres = ['hip', 'hop', 'hip-hop']
correct_genre = 'hiphop'

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [20]:
# Eliminar duplicados implícitos
df = replace_wrong_genres(wrong_genres, correct_genre)

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [21]:
# Comprobación de duplicados implícitos
print(df['genre'].sort_values().unique())

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

### Tus observaciones


`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`
>Se analizaron dumplicados explicitos y se eliminaron para no afectar nuestro analisis. Con los duplicados implicitos se detecto que habia generos repetidos pero con variaciones en su ortografia por lo que se abordaron con una funcion para reemplazarlos

## Prueba de hipótesis

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.

Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.

In [22]:
# Contar las canciones reproducidas en cada ciudad
print(df.groupby('city')['track'].count())

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


`Comenta tus observaciones aquí`
>Springfield tiene __42741__ canciones reproducidas y Shelbyville __18512__. Una diferencia bastante significativa de __24229__ canciones reproducidas

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.

In [23]:
# Calcular las canciones reproducidas en cada uno de los tres días
print(df.groupby('day')['track'].count())

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64


`Comenta tus observaciones aquí`
>Los viernes se reproducieron __21840__ canciones, los miercoles __21354__ y los miercoles __18059__. Los dias viernes es el dia de la semana que mas canciones se reproducen.

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [24]:
# Declara la función number_tracks() con dos parámetros: day= y city=
def number_tracks(day, city):
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    day_filter = df[df['day'] == day]
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    city_filter = day_filter[day_filter['city'] == city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    total_users = city_filter['user_id'].count()
    # Devolve el número de valores de la columna 'user_id'
    return total_users

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [25]:
# El número de canciones reproducidas en Springfield el lunes
print('El numero de canciones reproducidas en Springfield el lunes es:', number_tracks('Monday', 'Springfield'))

El numero de canciones reproducidas en Springfield el lunes es: 15740


In [26]:
# El número de canciones reproducidas en Shelbyville el lunes
print('El numero de canciones reproducidas en Shelbyville el lunes es:', number_tracks('Monday', 'Shelbyville'))

El numero de canciones reproducidas en Shelbyville el lunes es: 5614


In [27]:
# El número de canciones reproducidas en Springfield el miércoles
print('El numero de canciones reproducidas en Springfield el miercoles es:', number_tracks('Wednesday', 'Springfield'))

El numero de canciones reproducidas en Springfield el miercoles es: 11056


In [28]:
# El número de canciones reproducidas en Shelbyville el miércoles
print('El numero de canciones reproducidas en Shelbyville el miercoles es:', number_tracks('Wednesday', 'Shelbyville'))

El numero de canciones reproducidas en Shelbyville el miercoles es: 7003


In [29]:
# El número de canciones reproducidas en Springfield el viernes
print('El numero de canciones reproducidas en Springfield el viernes es:', number_tracks('Friday', 'Springfield'))

El numero de canciones reproducidas en Springfield el viernes es: 15945


In [30]:
# El número de canciones reproducidas en Shelbyville el viernes
print('El numero de canciones reproducidas en Shelbyville el viernes es:', number_tracks('Friday', 'Shelbyville'))

El numero de canciones reproducidas en Shelbyville el viernes es: 5895


**Conclusiones**

`Comenta si la hipótesis es correcta o se debe rechazar. Explica tu razonamiento.`
>Nuestra hipotesis es la siguiente: La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.

> Podemos decir que nuestra hipotesis es correcta ya que en ambas ciudades difiere la actividad segun el dia de las semana

# Conclusiones

In [31]:
# Numero de usuarios por cada ciudad
print(df.groupby('city')['user_id'].nunique())

city
Shelbyville    12423
Springfield    29358
Name: user_id, dtype: int64


`Resume aquí tus conclusiones sobre la hipótesis.`
>Se puede ver que en Springfield hay mas del doble de usuarios que en Shelbyville por lo que siempre veremos mas cantidad de reproducciones cualquier dia de la semana en Sprigfield, pero puedo ver que los dias "Lunes" y "Viernes" Springfield tiene una media mas alta de reproducciones de casi el 50% mas de usuarios pero el dia "Miercoles" Shelbyville aumenta muchisimo su media de reproducciones a compracion de Sprigfield

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.